google/cloud/firestore_v1/throttle.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2021 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+from typing import Callable, NoReturn, Optional
+
+
+def utcnow():
+    return datetime.datetime.utcnow()
+
+
+default_initial_tokens: int = 500
+default_phase_length: int = 60 * 5  # 5 minutes
+microseconds_per_second: int = 1000000
+
+
+class Throttle:
+    """Implements 5/5/5 ramp-up via Token Bucket algorithm."""
+
+    def __init__(
+        self,
+        initial_tokens: Optional[int] = default_initial_tokens,
+        phase_length: Optional[int] = default_phase_length,
+    ):
+        # Tracks the volume of operations during a given ramp-up phase.
+        self._operations_this_phase: int = 0
+
+        self._start: datetime.datetime = utcnow()
+        self._last_refill: datetime.datetime = self._start
+
+        # Current number of available operations. Decrements with every
+        # permitted request and refills over time.
+        self._available_tokens: int = initial_tokens
+
+        # Maximum size of the available operations. Can increase by 50%
+        # every [phase_length] number of seconds.
+        self._maximum_tokens: int = self._available_tokens
+
+        # Number of seconds after which the [_maximum_tokens] can increase by 50%.
+        self._phase_length: int = phase_length
+
+        # Tracks how many times the [_maximum_tokens] has increased by 50%.
+        self._phase: int = 0
+
+    def take_tokens(self, num: Optional[int] = 1) -> bool:
+        """Returns True if an operation is currently permitted or False if not."""
```

Comment:
what does it mean to be permitted?

Reply:
this docstring and type annotation is actually just outright stale. updating.